Fix persistent settings

RubenSandwich opened the pull request:
Review needed: @thomasguerena

RubenSandwich committed:
Fix persistent settings

thomasguerena commented on FFOSBR.js:
```diff
@@ -12,67 +12,78 @@
+    repeat: true
   };
 
+  // Load options if present
```
Not sure this needs to be wrapped in an ASCF. If this is just "constructor" code, it will be run automatically when a new instance of Settings is created. No need for the anon function.

thomasguerena commented on FFOSBR.js:
```diff
@@ -12,67 +12,78 @@
 
-
-    if (myDictionary.photos !== undefined && typeof myDictionary.photos === "boolean") {
+    if (myDictionary.photos !== undefined &&
```
All the if-statements can be reduced to "if (typeof myDictionary.photos === 'boolean') ..." because "undefined" is also a type returned by typeof. Meaning, there's no harm in calling typeof on an undefined variable and if its type is "boolean" then you know it isn't undefined anyway.

thomasguerena commented on testrunner/FFOSBR.js:
```diff
@@ -12,67 +12,78 @@
+    repeat: true
   };
 
+  // Load options if present
```
Any reason this "constructor" code in an ASCF? Typically, this should be in the body of your Settings class, similar to your "timeInMilliSec" and "options".

RubenSandwich replied:
The reason why this is a ASCF is because I didn't want to store 'retrievedOptions' in the object or call  'localStorage.getItem('ffosbrOptions')' twice.

thomasguerena replied:
Ok. I think I see the issue here. Just so you, none of the options are being saved to the Settings object. Meaning, "Settings.options" is undefined. I think that's also why there's confusion about needing the ASCF. Anything declared as "var foo" in a constructor only exists in the constructor. Anything declared as "this.foo" is attached to the object. This would explain why the prototypes had to be defined within the constructor, which are typically defined outside the constructor, otherwise they are redefined every time a new instance of the object is made. I'm going to do a quick refactor if that's ok with you. The changes are minor.

RyanLB replied:
Is the aforementioned refactor finished, or still in progress?

thomasguerena commented on testrunner/FFOSBR.js:
```diff
@@ -12,67 +12,78 @@
       return options;
     }
 
-
```
All the if-statements can be reduced to "if (typeof myDictionary.photos === 'boolean') ..." because "undefined" is also a type returned by typeof. Meaning, there's no harm in calling typeof on an undefined variable and if its type is "boolean" then you know it isn't undefined anyway.

thomasguerena commented:
I couldn't comment on the Settings module, so I left some comments in testrunner/FFOSBR. The if-statement changes are minimal, so I think I'll go ahead and make them myself. However, I'd like to know if there's a particular reason for the ASCF constructor before I make any changes to it.

thomasguerena committed:
Removed unnecessary checks in Settings module.

thomasguerena committed:
Refactored Settings module. Still need to rewrite the unit tests and bug scrub.

thomasguerena committed:
Added better single field validation to Settings module.

thomasguerena committed:
Added unit tests for Settings module refactor. Yet to be run or tested.

thomasguerena committed:
Refactored Settings module and added unit tests.

RyanLB commented:
How's this refactor coming? Does this need to be reviewed again?

RubenSandwich commented:
Yes you review it. Both Thomas and I have worked on this so it doesn't make sense for us to review it. But I just spoke with Thomas and he says it done.

RyanLB commented on FFOSBR.js:
```diff
@@ -12,67 +12,136 @@
+  var retrievedOptions = localStorage.getItem('ffosbrOptions');
+
+  if (retrievedOptions !== null) {
+    this.options = JSON.parse(retrievedOptions);
```
Should we be doing any sort of validation here to ensure that what we parsed was a valid options dictionary?

RyanLB commented on modules/settings.js:
```diff
@@ -13,63 +12,120 @@ function Settings() {
-    }
-  })();
+  if (retrievedOptions !== null) {
+    this.options = JSON.parse(retrievedOptions);
```
Do we need any sort of validation here to make sure what we parsed is a valid options object?

thomasguerena replied:
Good point. We could just pass the parsed object into the validate function. I'll add that.

RyanLB commented on modules/settings.js:
```diff
@@ -13,63 +12,120 @@ function Settings() {
+  return valid;
+};
+
+Settings.prototype.set = function(newOptions) {
```
Do we want a method that lets you add or update individual fields as well?

thomasguerena replied:
I can't think of a need for that, but if you can (or just want it) it would be very easy to add.

RyanLB replied:
My impression was that that was how you'd update settings through the UI. If I was wiring up a button to turn on photo backup, I was thinking the event handler would just call something like set('photos', true).

If we had another plan, though, that's fine – since we have to write the whole settings object to disk anyway, that layer of abstraction isn't strictly necessary.

RubenSandwich replied:
I agree with Ryan that it would be useful to do something like: ffosbr.settings.set({photos: true}); and not have it set all the other fields to undefined. Also why is ffosbr.settings.options public? It's confusing to me because we validate the data when using set but we also allow them to set it directly which completely bypasses the validate functionality.

thomasguerena replied:
Both settings.set('photos', true) and settings.set({photos: true}) already exist. See Settings.set().

thomasguerena replied:
Oh never mind. I was thinking of Settings.validate(). I'll add the same functionality to Settings.set().

RubenSandwich replied:
My issue has more to do with setting.options being public and editable therefore bypassing the nice validation method you wrote.

RyanLB commented on modules/settings.js:
```diff
@@ -13,63 +12,120 @@ function Settings() {
+    return console.log('Invalid config object', newOptions);
+  }
+
+  if (this.validate('photos', newOptions.photos)) {
```
I think if we leave potentialOptions undefined, it'll validate all of our options at once, meaning we don't need this giant block of if statements. Using this would remove the ability to skip over invalid components of the options object while still setting the valid ones. Is this something we want? Or should we assume that if a field in this object is invalid, the object as a whole shouldn't be trusted?

thomasguerena replied:
It probably makes more sense to just ignore any object which contains invalid fields. I'll switch it around. It also reduces code (although I could reduce the current code, anyway).

thomasguerena committed:
Added multiple options for Settings.set(). Added validation to local-storage load.

RyanLB commented on modules/settings.js:
```diff
@@ -13,63 +12,158 @@ function Settings() {
-    }
+  var valid = true;
+  var opts = null;
+  var validTypes = {
```
Should this be declared outside this function so it doesn't get re-initialized with every call to validate()?

RyanLB committed:
Merge branch 'master' into fixPersistentSettings

* master: (29 commits)
  Updating .gitignore so that it ignores bower_components in the top-level directory and the DemoUI directory (or directories for any other future test apps).
  checking for external now
  Fixed bug which wouldn't allow undefined directory parameter in Media.get().
  Media.remove() passing all unit tests.
  Media.put() tests passing.
  Fixed all tests and bugs found by those tests for Media.get().
  Fixed style errors reported by the build.
  Improved error detection and optional arguments in Media.get().
  Fixed numerous build and style errors.
  Media.get() now accepts a 'directory' parameter specifying a path in sdcard1 from which to fetch its files.
  Added tests for exceptions
  Added initial backup/restore/clean functionality. Needs testing.
  Update media.spec.js
  added comments
  can place photos in testphotos dir for testing.
  testing for photo backup
  ./photos breaks the tests
  Flushing out bugs while implementing unit tests. Unfinished.
  add functionality of convert mozContact to vcard format
  Added additional error checking and some external-sd-only functionality.
  ...

Conflicts:
	FFOSBR.js
	FFOSBR.min.js
	testrunner/FFOSBR.js
	testrunner/tests/settings.spec.js

RyanLB committed:
Did a bad job merging master into fixPersistentSettings

thomasguerena committed:
Added temporary fix for missing FFOSBR script in Settings unit tests. This will be refactored soon.

thomasguerena committed:
Centralized backup directory paths into the Settings module.

thomasguerena committed:
Removed unused variable from Restore.

thomasguerena committed:
Added initial code for Media.getFreeBytes(). Not quite working, yet.

RyanLB committed:
Make settings tests synchronous. Fix a small but serious error in settings retrieval.
• Simply clobbered values and reloaded settings manually rather than asynchronously removing/re-adding FFOSBR in settings.spec.js
• Moved a paren so that we didn't always pass a bool to settings.validate()

RyanLB committed:
Remove a reference to an undefined variable. This is the tiniest commit.

RyanLB committed:
Another undefined variable in the settings module. The second tiniest commit.

RyanLB commented:
Closing this pull request for now; @thomasguerena, reopen it when you're ready.
